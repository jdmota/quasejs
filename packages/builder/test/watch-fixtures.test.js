import index from "../dist";
import Reporter from "../dist/reporter";
import transformConfig from "./transform-config";
import { cleanText, compareFolders } from "./expect";

const fs = require( "fs-extra" );
const path = require( "path" );

const FIXTURES = path.resolve( "packages/builder/test/watch-fixtures" );
const folders = fs.readdirSync( FIXTURES );

class TestReporter extends Reporter {
  constructor( opts, builder ) {
    super( opts, builder );
    this.output = "";
  }
  _log( message ) {
    this.output += cleanText( message ) + "\n";
  }
}

describe( "watcher", () => {

  folders.forEach( folder => {

    if (
      folder === "__dev__" ||
      // FIXME
      folder === "watches-other-extensions"
    ) {
      return;
    }

    jest.setTimeout( 60000 );

    it( `Fixture: watch-fixtures/${folder}`, async() => {

      const fixturePath = path.resolve( FIXTURES, folder );
      const filesPath = path.resolve( fixturePath, "files" );
      const workingPath = path.resolve( fixturePath, "working" );
      const expectedPath = path.resolve( fixturePath, "expected" );
      const actualPath = path.resolve( fixturePath, "actual" );

      await fs.emptyDir( actualPath );
      await fs.emptyDir( workingPath );
      await fs.copy( filesPath, workingPath );

      let config;
      try {
        config = require( path.resolve( fixturePath, "config.js" ) );
      } catch ( err ) {
        config = {};
      }

      config.cwd = fixturePath;
      config.entries = config.entries || [ "index.js" ];
      config.context = config.context || "working";
      config.dest = config.dest || "actual";
      config.watch = true;
      config.watchOptions = {
        aggregateTimeout: 1
      };

      config.reporter = [
        TestReporter,
        {
          logLevel: 5,
          color: false,
          isTest: true
        }
      ];

      config = transformConfig( config, fixturePath );

      const { reporter, builder } = index( config, true );
      const { watcher } = builder;

      function update( file, type ) {
        file = path.sep === "\\" ? file.toLowerCase() : file;
        if ( watcher.watchedFiles().has( file ) ) {
          watcher._onUpdate( file, type );
          return true;
        }
        return false;
      }

      let resolve;
      const promise = new Promise( a => {
        resolve = a;
      } );

      const operations = require( path.resolve( fixturePath, "operations.js" ) );
      let i = 0;

      async function next() {

        if ( i >= operations.length ) {
          builder.stop();
          resolve();
          return;
        }

        let [ action, arg1, arg2 ] = operations[ i ];
        arg1 = path.resolve( workingPath, arg1 );

        switch ( action ) {
          case "newFile":
            await fs.writeFile( arg1, arg2 );
            if ( update( arg1, "added" ) ) {
              watcher.queueBuild();
            }
            break;
          case "writeFile":
            await fs.writeFile( arg1, arg2 );
            if ( update( arg1, "changed" ) ) {
              watcher.queueBuild();
            }
            break;
          case "ensureDir":
            await fs.ensureDir( arg1 );
            break;
          case "rename":
            arg2 = path.resolve( workingPath, arg2 );
            await fs.rename( arg1, arg2 );
            if ( update( arg1, "removed" ) || update( arg2, "added" ) ) {
              watcher.queueBuild();
            }
            break;
          case "remove":
            await fs.remove( arg1 );
            if ( update( arg1, "removed" ) ) {
              watcher.queueBuild();
            }
            break;
          default:
            throw new Error( `Invalid action: ${action}` );
        }

        i++;
        watcher.currentBuild.then( next );
      }

      watcher.currentBuild.then( next );

      await promise;

      expect( reporter.output ).toMatchSnapshot();

      await compareFolders( actualPath, expectedPath );

    } );

  } );

} );
