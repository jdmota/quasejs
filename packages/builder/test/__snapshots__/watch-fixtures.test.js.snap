// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`watcher Fixture: watch-fixtures/basic 1`] = `
"Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.33 kB   

√  Built!
Watching 8 files...
"
`;

exports[`watcher Fixture: watch-fixtures/custom-code-frame 1`] = `
"Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:14: Could not resolve ./a
Watching 11 files...
i  File working/a.js was added.
Warming up...
Building...
Checking...
×  index.js[quase_builder_js_transformer]:1:7: ./a exports nothing. Looking for default
Watching 11 files...
i  File working/a.js was changed.
Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.43 kB   

√  Built!
Watching 11 files...
i  File working/a.js was removed.
Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:14: Could not resolve ./a
Watching 11 files...
"
`;

exports[`watcher Fixture: watch-fixtures/performance-hints 1`] = `
"Warming up...
Building...
Checking...
Computing graph...
Creating files...

          index.js      356 B   > 100 B [performance!] 
 [entry]  index.html  5.43 kB   > 10 B [performance!]  

√  Built!
Watching 14 files...
"
`;

exports[`watcher Fixture: watch-fixtures/resolve-recovery 1`] = `
"Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.33 kB   

√  Built!
Watching 8 files...
i  File working/x.js was removed.
Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:14: Could not resolve ./x
Watching 11 files...
i  File working/x/index.js was changed.
Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.35 kB   

√  Built!
Watching 11 files...
"
`;

exports[`watcher Fixture: watch-fixtures/script-inline-in-html 1`] = `
"Warming up...
Building...
×  index.html[quase_builder_html_transformer](0)[quase_builder_js_transformer]:2:24: Could not resolve ./a.js
Watching 11 files...
i  File working/a.js was added.
Warming up...
Building...
Checking...
×  index.html[quase_builder_html_transformer](0)[quase_builder_js_transformer]:2:15: ./a.js exports nothing. Looking for a
Watching 11 files...
i  File working/a.js was changed.
Warming up...
Building...
Checking...
Computing graph...
Creating files...

          index.js      450 B   
 [entry]  index.html  5.54 kB   

√  Built!
Watching 11 files...
"
`;

exports[`watcher Fixture: watch-fixtures/watches-missing-entry 1`] = `
"Warming up...
Building...
×  Could not find packages/builder/test/watch-fixtures/watches-missing-entry/working/index.js
Watching 1 files...
i  File working/index.js was added.
Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.21 kB   

√  Built!
Watching 1 files...
"
`;

exports[`watcher Fixture: watch-fixtures/watches-missing-file 1`] = `
"Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:14: Could not resolve ./a
Watching 11 files...
i  File working/a.js was added.
Warming up...
Building...
Checking...
×  index.js[quase_builder_js_transformer]:1:7: ./a exports nothing. Looking for default
Watching 11 files...
i  File working/a.js was changed.
Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.33 kB   

√  Built!
Watching 11 files...
i  File working/a.js was removed.
Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:14: Could not resolve ./a
Watching 11 files...
"
`;

exports[`watcher Fixture: watch-fixtures/watches-modules-in-cache 1`] = `
"Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.33 kB   

√  Built!
Watching 8 files...
i  File working/index.js was changed.
Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:18: Could not resolve ./stuff
Watching 13 files...
i  File working/a.js was changed.
Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:18: Could not resolve ./stuff
Watching 13 files...
i  File working/index.js was changed.
Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.33 kB   

√  Built!
Watching 13 files...
"
`;

exports[`watcher Fixture: watch-fixtures/watches-rename 1`] = `
"Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.33 kB   

√  Built!
Watching 8 files...
i  File working/a.js was removed.
Warming up...
Building...
×  index.js[quase_builder_js_transformer]:1:14: Could not resolve ./a
Watching 11 files...
i  File working/index.js was changed.
Warming up...
Building...
Checking...
Computing graph...
Creating files...

 [entry]  index.js  5.33 kB   

√  Built!
Watching 13 files...
"
`;
